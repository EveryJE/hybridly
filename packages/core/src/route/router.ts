import { stringify } from 'qs'
import { getInternalRouterContext } from '../context'
import { Route } from './route'
import type { RouteName, RouteParameters, RoutingConfiguration } from './types'

/**
 * A collection of Laravel routes. This class constitutes Ziggy's main API.
 */
export class Router<T extends RouteName> extends String {
	public route: Route
	private parameters!: RouteParameters<T>
	private routing: RoutingConfiguration

	constructor(
		name: T,
		parameters?: RouteParameters<T>,
		absolute: boolean = true,
	) {
		super()
		const context = getInternalRouterContext()
		this.route = new Route(name, absolute)
		this.routing = context.routing!
		this.setParameters(parameters)
	}

	/**
	 * Gets the compiled URL string for the current route and parameters.
	 *
	 * @example
	 * // with 'posts.show' route 'posts/{post}'
	 * (new Router('posts.show', 1)).toString(); // 'https://ziggy.dev/posts/1'
	 */
	toString(): string {
		// Get parameters that don't correspond to any route segments to append them to the query
		const unhandled = Object.keys(this.parameters)
			.filter((key) => !this.route.parameterSegments.some(({ name }) => name === key))
			.filter((key) => key !== '_query')
			.reduce((result, current) => ({ ...result, [current]: this.parameters[current] }), {})

		return this.route.compile(this.parameters) + stringify({ ...unhandled, ...this.parameters._query }, {
			addQueryPrefix: true,
			arrayFormat: 'indices',
			encodeValuesOnly: true,
			skipNulls: true,
			encoder: (value, encoder) => typeof value === 'boolean'
				? Number(value).toString()
				: encoder(value),
		})
	}

	/**
	 * Checks whether the given route exists.
	 */
	static has(name: string): boolean {
		try {
			Route.getDefinition(name)
			return true
		} catch {
			return false
		}
	}

	/**
	 * Parse Laravel-style route parameters of any type into a normalized object.
	 *
	 * @example
	 * // with route parameter names 'event' and 'venue'
	 * _parse(1); // { event: 1 }
	 * _parse({ event: 2, venue: 3 }); // { event: 2, venue: 3 }
	 * _parse(['Taylor', 'Matt']); // { event: 'Taylor', venue: 'Matt' }
	 * _parse([4, { uuid: 56789, name: 'Grand Canyon' }]); // { event: 4, venue: 56789 }
	 */
	private setParameters(parameters?: RouteParameters<T>) {
		this.parameters = parameters ?? {}

		// If `params` is a string or integer, wrap it in an array
		this.parameters = ['string', 'number'].includes(typeof this.parameters) ? [this.parameters] : this.parameters

		// Separate segments with and without defaults, and fill in the default values
		const segments = this.route.parameterSegments.filter(({ name }) => !this.routing.defaults[name])

		if (Array.isArray(this.parameters)) {
			// If the parameters are an array they have to be in order, so we can transform them into
			// an object by keying them with the template segment names in the order they appear
			this.parameters = this.parameters.reduce((result, current, i) => segments[i]
				? ({ ...result, [segments[i].name]: current })
				: typeof current === 'object'
					? ({ ...result, ...current })
					: ({ ...result, [current]: '' }), {})
		} else if (
			segments.length === 1
            && !this.parameters[segments[0].name]
            && (Reflect.has(this.parameters, Object.values(this.route.definition.bindings)[0]) || Reflect.has(this.parameters, 'id'))
		) {
			// If there is only one template segment and `this.parameters` is an object, that object is
			// ambiguousâ€”it could contain the parameter key and value, or it could be an object
			// representing just the value (e.g. a model); we can inspect it to find out, and
			// if it's just the parameter value, we can wrap it in an object with its key
			this.parameters = { [segments[0].name]: this.parameters }
		}

		this.parameters = {
			...this.getDefaults(),
			...this.substituteBindings(),
		}
	}

	/**
	 * Populates default parameters.
	 */
	private getDefaults() {
		return this.route.parameterSegments
			.filter(({ name }) => this.routing.defaults[name])
			.reduce((result, { name }) => ({ ...result, [name]: this.routing.defaults[name] }), {})
	}

	/**
     * Substitute Laravel route model bindings in the given parameters.
     *
     * @example
     * _substituteBindings({ post: { id: 4, slug: 'hello-world', title: 'Hello, world!' } }, { bindings: { post: 'slug' } }); // { post: 'hello-world' }
     */
	private substituteBindings() {
		return Object.entries(this.parameters).reduce((result, [key, value]) => {
			// If the value isn't an object, or if the key isn't a named route parameter,
			// there's nothing to substitute so we return it as-is
			if (!value || typeof value !== 'object' || Array.isArray(value) || !this.route.parameterSegments.some(({ name }) => name === key)) {
				return { ...result, [key]: value }
			}

			if (!Reflect.has(value, this.route.definition.bindings[key])) {
				// As a fallback, we still accept an 'id' key not explicitly registered as a binding
				if (Reflect.has(value, 'id')) {
					this.route.definition.bindings[key] = 'id'
				} else {
					throw new Error(`Router error: object passed as [${key}] parameter is missing route model binding key [${this.route.definition.bindings?.[key]}].`)
				}
			}

			return { ...result, [key]: value[this.route.definition.bindings[key]] }
		}, {})
	}

	valueOf() {
		return this.toString()
	}
}
