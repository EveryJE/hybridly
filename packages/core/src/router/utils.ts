export const STORAGE_EXTERNAL_KEY = 'sleightful:external'

/**
 * Checks if the current visit was made by going back or forward.
 */
export function isBackForwardVisit(): boolean {
	if (!window.history.state) {
		return false
	}

	return (window.performance?.getEntriesByType('navigation').at(0) as PerformanceNavigationTiming)?.type === 'back_forward'
}

/**
 * Checks if the current visist points to an external location.
 */
export function isExternalVisit(): boolean {
	try {
		return window.sessionStorage.getItem(STORAGE_EXTERNAL_KEY) !== null
	} catch {}

	return false
}

/**
 * Gets the current visit type.
 */
export function getCurrentVisitType(): 'back_forward' | 'external' | 'sleightful' {
	if (isBackForwardVisit()) {
		return 'back_forward'
	}

	if (isExternalVisit()) {
		return 'external'
	}

	return 'sleightful'
}

/**
 * Convers the given string or URL to an URL.
 */
export function makeUrl(href: string | URL): URL {
	return new URL(href.toString(), window.location.toString())
}
