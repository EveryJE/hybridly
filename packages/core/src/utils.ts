import makeDebugger from 'debug'

export const debug = {
	router: makeDebugger('sleightful:router'),
	history: makeDebugger('sleightful:history'),
	url: makeDebugger('sleightful:url'),
	context: makeDebugger('sleightful:context'),
	external: makeDebugger('sleightful:external'),
	scroll: makeDebugger('sleightful:scroll'),
	event: makeDebugger('sleightful:event'),
	adapter: (name: string, ...args: any[]) => makeDebugger('sleightful:adapter').extend(name)(args.shift(), ...args),
}

export function match<TValue extends string | number = string, TReturnValue = unknown>(
	value: TValue,
	lookup: Record<TValue | 'default', TReturnValue | ((...args: any[]) => TReturnValue)>,
	...args: any[]
): TReturnValue | Promise<TReturnValue> {
	if (value in lookup || 'default' in lookup) {
		const returnValue = value in lookup
			? lookup[value]
			: lookup.default

		return typeof returnValue === 'function' ? returnValue(...args) : returnValue
	}

	const handlers = Object.keys(lookup)
		.map((key) => `"${key}"`)
		.join(', ')

	const error = new Error(`Tried to handle "${value}" but there is no handler defined. Only defined handlers are: ${handlers}.`)

	if (Error.captureStackTrace) {
		Error.captureStackTrace(error, match)
	}

	throw error
}

export function debounce<F extends(...params: any[]) => ReturnType<F>>(fn: F, delay: number): F {
	let timeoutID: NodeJS.Timeout
	return function(...args: unknown[]) {
		clearTimeout(timeoutID)
		timeoutID = setTimeout(() => fn(args), delay)
	} as F
}

export function value<T>(value: T | (() => T)): T {
	if (typeof value === 'function') {
		return (value as any)?.() as T
	}

	return value
}

export function when<T, D>(condition: any, data: T, _default?: D): T | D | undefined {
	if (!condition) {
		return _default
	}

	return data
}
