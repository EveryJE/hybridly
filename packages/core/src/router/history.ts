import { SCROLL_REGION_ATTRIBUTE } from '../constants'
import { debounce, debug } from '../utils'
import { navigate } from './router'
import { RouterContext } from './context'
import { saveScrollPositions } from './scroll'
import { makeUrl } from './url'

type SerializedContext = Omit<RouterContext, 'adapter' | 'events'>

/** Puts the given context into the history state. */
export function setHistoryState(context: RouterContext, options: HistoryOptions = {}) {
	if (!window?.history) {
		throw new Error('The history API is not available, so Sleightful cannot operate.')
	}

	const method = options.replace
		? 'replaceState'
		: 'pushState'

	debug.history('Setting history state:', { method, context })

	try {
		window.history[method](serializeContext(context), '', context.url)
	} catch (error) {
		console.error('Sleightful could not save its current state in the history. This is most likely due to a property being non-serializable, such as a proxy or a reference.')
		throw error
	}
}

/** Gets the current history state if it exists. */
export function getHistoryState<T extends keyof HistoryState | undefined>(
	key?: T,
): T extends keyof HistoryState ? HistoryState[T] : HistoryState | undefined {
	if (window?.history.state) {
		return key
			? window.history.state[key]
			: window.history.state
	}

	return key
		? undefined as any
		: { state: {} }
}

/** Register history-related event listeneners. */
export async function registerEventListeners(context: RouterContext) {
	debug.history('Registering [popstate] and [scroll] event listeners.')

	// Popstate is for catching back and forward navigations. We want
	// to imitate native browser behavior while keeping the SPA feeling.
	window?.addEventListener('popstate', async(event) => {
		debug.history('Navigation detected (popstate event). State:', { state: event.state })

		// If there is no state in this history entry, we come from the user
		// replacing the URL manually. In this case, we want to restore everything
		// like it was. We need to copy the hash if any and restore the scroll positions.
		if (!event.state) {
			debug.history('There is no state. Adding hash if any and restoring scroll positions.')

			return await navigate(context, {
				payload: {
					...context,
					url: makeUrl(context.url, { hash: window.location.hash }).toString(),
				},
				preserveScroll: true,
				preserveState: true,
				replace: true,
			})
		}

		// If the history entry has been sleightfully tempered with, we want
		// to use it. We swap the components accordingly.
		await navigate(context, {
			payload: event.state,
			preserveScroll: true,
			preserveState: false,
			updateHistoryState: false,
		})
	})

	// On scroll, we want to save the positions of all scrollbars.
	// This is needed in order to restore them upon navigation.
	window?.addEventListener('scroll', (event) => debounce(() => {
		if ((event?.target as Element)?.hasAttribute?.(SCROLL_REGION_ATTRIBUTE)) {
			saveScrollPositions(context)
		}
	}, 100), true)
}

/** Checks if the current visit was made by going back or forward. */
export function isBackForwardVisit(): boolean {
	if (!window.history.state) {
		return false
	}

	return (window.performance?.getEntriesByType('navigation').at(0) as PerformanceNavigationTiming)?.type === 'back_forward'
}

/** Handles a visit which was going back or forward. */
export async function handleBackForwardVisit(context: RouterContext): Promise<void> {
	debug.router('Handling a back/forward visit.')
	window.history.state.version = context.version

	await navigate(context, {
		preserveScroll: true,
		preserveState: true,
	})
}

/** Serializes the context so it can be written to the history state. */
export function serializeContext(context: RouterContext): SerializedContext {
	return {
		url: context.url,
		version: context.version,
		view: context.view,
		dialog: context.dialog,
		scrollRegions: context.scrollRegions,
	}
}

export interface HistoryOptions {
	/**
	 * Replace the current history entry instead of pushing a new entry.
	 * @default false
	 */
	replace?: boolean
}

export interface HistoryState {
	state: any
}
