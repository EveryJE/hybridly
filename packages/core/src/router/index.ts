import type { DialogVisit, ExternalVisitOptions, NavigationOptions, ResolveComponent, RouterOptions, StatefulVisit, SwapDialog, SwapView, Visit } from '../types'
import { match } from '../utils'
import { getCurrentVisitType, makeUrl, STORAGE_EXTERNAL_KEY } from './utils'

/**
 * Initializes state for a visit.
 */
function statefulVisit(visit: Visit, state: Partial<StatefulVisit> = {}): StatefulVisit {
	return {
		visit,
		id: '',
		scrollPositions: [],
		state: {},
		...state,
	}
}

export class Router {
	public id: string
	protected resolveComponent: ResolveComponent
	protected swapView: SwapView
	protected swapDialog: SwapDialog
	protected currentState: StatefulVisit

	constructor(options: RouterOptions) {
		this.id = ''
		this.swapView = options.swap.view
		this.swapDialog = options.swap.dialog
		this.resolveComponent = options.resolve
		this.currentState = statefulVisit(options.visit)
		// this.events.emit('navigate', this.currentState.visit.view)
	}

	async initialize() {
		await match(getCurrentVisitType(), {
			back_forward: async() => await this.handleBackForwardVisit(),
			external: async() => await this.handleExternalVisit(),
			sleightful: async() => await this.handleVisit(),
		})
	}

	/*
	|--------------------------------------------------------------------------
	| Visit handling
	|--------------------------------------------------------------------------
	*/

	private async handleExternalVisit(): Promise<void> {
		const options = JSON.parse(window.sessionStorage.getItem(STORAGE_EXTERNAL_KEY) || '{}') as ExternalVisitOptions
		window.sessionStorage.removeItem(STORAGE_EXTERNAL_KEY)

		await this.navigate({
			visit: this.currentState.visit,
			preserveScroll: options.preserveScroll,
			preserveState: true,
		})
	}

	private async handleBackForwardVisit(): Promise<void> {
		window.history.state.version = this.currentState.visit.version

		await this.navigate({
			visit: window.history.state,
			preserveScroll: true,
			preserveState: true,
		})

		this.restoreScrollPositions()
	}

	private async handleVisit(): Promise<void> {
		await this.navigate({
			visit: this.currentState.visit,
			preserveState: true,
		})
	}

	/*
	|--------------------------------------------------------------------------
	| Core navigation
	|--------------------------------------------------------------------------
	*/

	async navigate(options: NavigationOptions): Promise<void> {
		const visit = options.visit as DialogVisit

		// TODO: preserve hash
		// window.location.hash - not sure where the logic should belong

		const shouldReplace = options.replace || makeUrl(visit.dialog?.url ?? visit.view?.url).href === window.location.href
		const pageComponent = await this.resolveComponent(visit.view.name)
		const dialogComponent = options.visit.type === 'dialog'
			? await this.resolveComponent(visit.dialog.name)
			: undefined

		// The view state needs to be preserved if we have a dialog.
		// Otherwise, we preserve it only when specified.
		await this.swapView({
			view: visit.view,
			component: pageComponent,
			preserveState: dialogComponent ? true : options.preserveState,
		})

		await this.swapDialog({
			component: dialogComponent,
			preserveState: options.preserveState,
		})

		this.setHistoryState(options.visit, shouldReplace)

		if (options.preserveScroll) {
			this.restoreScrollPositions()
		}

		if (!shouldReplace) {
			// this.events.emit('navigate', this)
		}
	}

	// private setupEventListeners(): void {
	// 	window.addEventListener('popstate', this.handlePopstateEvent.bind(this))
	// 	document.addEventListener('scroll', debounce(this.handleScrollEvent.bind(this), 100), true)
	// }

	/*
	|--------------------------------------------------------------------------
	| Scroll regions
	|--------------------------------------------------------------------------
	*/

	/**
	 * Gets the elements which scroll positions should be preserved.
	 */
	scrollRegions() {
		return document.querySelectorAll('[scroll-region]')
	}

	async restoreScrollPositions(): Promise<void> {
		const positions = this.scrollRegions()

		if (!this.currentState.scrollPositions) {
			return
		}

		const timer = setInterval(() => {
			// TODO: find a better way to ensure the page component is fully loaded
			if (this.currentState.scrollPositions.length !== positions.length) {
				console.log('[sleightful] waiting for component to load')
				return
			}

			clearInterval(timer)
			positions.forEach((el: Element, i) => {
				el.scrollTop = this.currentState.scrollPositions.at(i)?.top ?? el.scrollTop
				el.scrollLeft = this.currentState.scrollPositions.at(i)?.left ?? el.scrollLeft
			})
		}, 10)
	}

	/*
	|--------------------------------------------------------------------------
	| History state
	|--------------------------------------------------------------------------
	*/

	/**
	 * Updates the history state for the given view.
	 */
	protected setHistoryState(visit: Visit, replace: boolean): void {
		// TODO - maybe try to move
		this.currentState.visit = visit

		window.history[replace ? 'replaceState' : 'pushState'](
			visit, // data for this history entry
			'', // this is unused, great api design haha
			visit.type === 'dialog' ? visit.dialog.url : visit.view.url, // url of this history entry
		)
	}
}

/**
 * Initializes the sleightful router.
 */
export async function initialize(options: RouterOptions): Promise<Router> {
	const router = new Router(options)
	await router.initialize()

	return router
}
