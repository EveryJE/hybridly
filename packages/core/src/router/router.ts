import axios, { AxiosResponse } from 'axios'
import qs from 'qs'
import { showResponseErrorModal, match, merge, when, debug, random } from '@hybridly/utils'
import { ERROR_BAG_HEADER, EXCEPT_DATA_HEADER, EXTERNAL_VISIT_HEADER, ONLY_DATA_HEADER, PARTIAL_COMPONENT_HEADER, HYBRIDLY_HEADER, VERSION_HEADER } from '../constants'
import { NotAHybridlyResponseError, VisitCancelledError } from '../errors'
import { getRouterContext, initializeContext, payloadFromContext, InternalRouterContext, RouterContextOptions, setContext } from '../context'
import { triggerEvent } from '../events'
import { handleExternalVisit, isExternalResponse, isExternalVisit, performExternalVisit } from '../external'
import { resetScrollPositions, restoreScrollPositions, saveScrollPositions } from '../scroll'
import { fillHash, makeUrl, normalizeUrl, sameUrls, UrlResolvable } from '../url'
import { setHistoryState, isBackForwardVisit, handleBackForwardVisit, registerEventListeners, getHistoryState, getKeyFromHistory, remember } from './history'
import type { ConditionalNavigationOption, Errors, LocalVisitOptions, NavigationOptions, Router, VisitOptions, VisitPayload, VisitResponse } from './types'

/**
 * The hybridly router.
 * This is the core function that you can use to navigate in
 * your application. Make sure the routes you call return a
 * hybridly response, otherwise you need to call `external`.
 *
 * @example
 * router.get('/posts/edit', { post })
 */
export const router: Router = {
	abort: async() => getRouterContext().activeVisit?.controller.abort(),
	active: () => !!getRouterContext().activeVisit,
	// unstack: () => {},
	visit: async(options) => await visit(options),
	reload: async(options) => await visit({ preserveScroll: true, preserveState: true, ...options }),
	get: async(url, options = {}) => await visit({ ...options, url, method: 'GET' }),
	post: async(url, options = {}) => await visit({ preserveState: true, ...options, url, method: 'POST' }),
	put: async(url, options = {}) => await visit({ preserveState: true, ...options, url, method: 'PUT' }),
	patch: async(url, options = {}) => await visit({ preserveState: true, ...options, url, method: 'PATCH' }),
	delete: async(url, options = {}) => await visit({ preserveState: true, ...options, url, method: 'DELETE' }),
	local: async(url, options) => await performLocalComponentVisit(url, options),
	external: (url, data = {}) => performLocalExternalVisit(url, data),
	history: {
		get: (key) => getKeyFromHistory(key),
		remember: (key, value) => remember(key, value),
	},
}

/** Creates the hybridly router. */
export async function createRouter(options: RouterContextOptions): Promise<InternalRouterContext> {
	initializeContext(options)

	return await initializeRouter()
}

/** Performs every action necessary to make a hybridly visit. */
export async function visit(options: VisitOptions): Promise<VisitResponse> {
	const visitId = random()
	const context = getRouterContext()
	debug.router('Making a visit:', { context, options, visitId })

	try {
		// Abort any active visit.
		if (context.activeVisit) {
			debug.router('Aborting current visit.', context.activeVisit)
			context.activeVisit?.controller.abort()
		}

		// Before anything else, we fire the "before" event to make sure
		// there was no user-specified handler returning "false".
		if (!(await triggerEvent('before', options, options.events?.before))) {
			debug.router('"before" event returned false, aborting the visit.')
			throw new VisitCancelledError('The visit was cancelled by the "before" event.')
		}

		// Before making the visit, we need to make sure the scroll positions are
		// saved, so we can restore them later.
		saveScrollPositions()

		// If the URL has transformation options, apply them before using the URL.
		if (options.url && options.transformUrl) {
			options.url = makeUrl(options.url, options.transformUrl)
		}

		// A visit is being made, we need to add it to the context so it
		// can be reused later on.
		setContext({
			activeVisit: {
				id: visitId,
				url: makeUrl(options.url ?? context.url),
				controller: new AbortController(),
				options,
			},
		})

		triggerEvent('start', context, options.events?.start)
		debug.router('Making request with axios.')

		const response = await axios.request({
			url: context.activeVisit!.url.toString(),
			method: options.method ?? 'GET',
			data: options.method === 'GET' ? {} : options.data,
			params: options.method === 'GET' ? options.data : {},
			signal: context.activeVisit!.controller.signal,
			headers: {
				...options.headers,
				...when(options.only?.length || options.except?.length, {
					[PARTIAL_COMPONENT_HEADER]: context.view.name,
					...when(options.only, { [ONLY_DATA_HEADER]: JSON.stringify(options.only) }, {}),
					...when(options.except, { [EXCEPT_DATA_HEADER]: JSON.stringify(options.except) }, {}),
				}, {}),
				...when(options.errorBag, { [ERROR_BAG_HEADER]: options.errorBag }, {}),
				...when(context.version, { [VERSION_HEADER]: context.version }, {}),
				// 'X-Hybridly-Context': this.currentState.visit.context,
				[HYBRIDLY_HEADER]: true,
				'X-Requested-With': 'XMLHttpRequest',
				'Accept': 'text/html, application/xhtml+xml',
			},
			validateStatus: () => true,
			onUploadProgress: (event: ProgressEvent) => {
				triggerEvent('progress', {
					event,
					percentage: Math.round(event.loaded / event.total * 100),
				}, options.events?.progress)
			},
		})

		triggerEvent('data', response, options.events?.data)

		// An external response is a hybridly response that wants a full page
		// load to a requested URL. It may be the same URL, in which case a
		// full page refresh will be performed.
		if (isExternalResponse(response)) {
			debug.router('The response is explicitely external.')
			await performExternalVisit({
				url: fillHash(context.activeVisit!.url, response.headers[EXTERNAL_VISIT_HEADER]),
				preserveScroll: options.preserveScroll === true,
			})

			return { response }
		}

		// An invalid response is a response that do not declare itself via
		// the protocole header.
		// In such cases, we want to throw to handler it later.
		if (!isHybridlyResponse(response)) {
			throw new NotAHybridlyResponseError(response)
		}

		// At this point, we know the response respects the hybridly protocol.
		debug.router('The response respects the hybridly protocol.')
		const payload = response.data as VisitPayload

		// If the visit was asking for specific properties, we ensure that the
		// new request object contains the properties of the current view context,
		// because the back-end sent back only the required properties.
		if ((options.only?.length ?? options.except?.length) && payload.view.name === context.view.name) {
			debug.router(`Merging ${options.only ? '"only"' : '"except"'} properties.`, payload.view.properties)
			payload.view.properties = merge(context.view.properties, payload.view.properties)
			debug.router('Merged properties:', payload.view.properties)
		}

		// If everything was according to the plan, we can make our navigation and
		// update the context. Underlying adapters get the updated data.
		await navigate({
			payload: {
				...payload,
				url: fillHash(context.activeVisit!.url, payload.url),
			},
			preserveScroll: options.preserveScroll === true,
			preserveState: options.preserveState,
			preserveUrl: options.preserveUrl,
			replace: options.replace === true || sameUrls(payload.url, window.location.href) || options.preserveUrl,
		})

		// If the new view's properties has errors, userland expects an event
		// with said errors to be emitted. However, errors can be scoped with
		// an error bag, and if the given error bag is missing, the event data
		// will be empty.
		if (Object.keys(context.view.properties.errors ?? {}).length > 0) {
			const errors = (() => {
				if (options.errorBag && typeof context.view.properties.errors === 'object') {
					return (context.view.properties.errors as any)[options.errorBag] ?? {}
				}

				return context.view.properties.errors
			})() as Errors

			debug.router('The request returned validation errors.', errors)
			triggerEvent('error', errors, options.events?.error)
			setContext({
				activeVisit: {
					...context.activeVisit as any,
					status: 'error',
				},
			})
		} else {
			triggerEvent('success', payload, options.events?.success)
			setContext({
				activeVisit: {
					...context.activeVisit as any,
					status: 'success',
				},
			})
		}

		return { response }
	//
	} catch (error: any) {
		match(error.constructor.name, {
			AbortError: () => {
				debug.router('The request was cancelled.', error)
				triggerEvent('abort', context, options.events?.abort)
			},
			NotAHybridlyResponseError: () => {
				debug.router('The request was not hybridly.')
				triggerEvent('invalid', error, options.events?.invalid)
				showResponseErrorModal(error.response.data)
			},
			default: () => {
				debug.router('An unknown error occured.', error)
				triggerEvent('exception', error, options.events?.exception)
			},
		})

		console.error(error)
		triggerEvent('fail', context, options.events?.fail)

		return {
			error: {
				type: error.constructor.name,
				actual: error,
			},
		}
	} finally {
		debug.router('Ending visit.')
		triggerEvent('after', context, options.events?.after)

		// If the visit is pending, it means another one has started,
		// and if we clean it up, it could lead to issues after the request is done.
		if (context.activeVisit?.id === visitId) {
			setContext({ activeVisit: undefined })
		}
	}
}

/** Checks if the response contains a hybridly header. */
export function isHybridlyResponse(response: AxiosResponse): boolean {
	return !!response?.headers[HYBRIDLY_HEADER]
}

/**
 * Makes an internal navigation that swaps the view and updates the context.
 * @internal This function is meant to be used internally.
 */
export async function navigate(options: NavigationOptions) {
	const context = getRouterContext()
	debug.router('Making an internal navigation:', { context, options })

	// If no request was given, we use the current context instead.
	options.payload ??= payloadFromContext()

	const evaluateConditionalOption = (option?: ConditionalNavigationOption) => typeof option === 'function'
		? option(options.payload!)
		: option

	const shouldPreserveState = evaluateConditionalOption(options.preserveState)
	const shouldPreserveScroll = evaluateConditionalOption(options.preserveScroll)
	const shouldReplaceHistory = evaluateConditionalOption(options.replace)
	const shouldReplaceUrl = evaluateConditionalOption(options.preserveUrl)

	// If the visit was asking to preserve the current state, we also need to
	// update the context's state from the history state.
	if (shouldPreserveState && getHistoryState() && options.payload.view.name === context.view.name) {
		setContext({ state: getHistoryState() })
	}

	// If the visit required the URL to be preserved, we skip its update
	// by replacing the payload URL with the current context URL.
	if (shouldReplaceUrl) {
		debug.router(`Preserving the current URL (${context.url}) instead of navigating to ${options.payload.url}`)
		options.payload!.url = context.url
	}

	// We merge the new request into the current context. That will replace
	// view, dialog, url and version, so the context is in sync with the
	// navigation that took place. We do not propagate the context
	// changes so adapters don't update props before the view.
	// We also reset the state so the state from one page
	// is not merged with the state from another one.
	setContext({
		...options.payload,
		state: {},
	}, { propagate: false })

	// History state must be updated to preserve the expected, native browser behavior.
	// However, in some cases, we just want to swap the views without making an
	// actual navigation. Additionally, we don't want to actually push a new state
	// when navigating to the same URL.
	if (options.updateHistoryState !== false) {
		debug.router(`Target URL is ${context.url}, current window URL is ${window.location.href}.`, { shouldReplaceHistory })
		setHistoryState({ replace: shouldReplaceHistory })
	}

	// Then, we swap the view.
	const viewComponent = await context.adapter.resolveComponent(context.view.name)
	debug.router(`Component [${context.view.name}] resolved to:`, viewComponent)
	await context.adapter.swapView({
		component: viewComponent,
		preserveState: shouldPreserveState,
	})

	// We then replace the dialog if a new one is given.
	if (context.dialog) {
		const dialogComponent = await context.adapter.resolveComponent(context.dialog.name)
		debug.router(`Dialog [${context.view.name}] resolved to:`, dialogComponent)
		await context.adapter.swapDialog({
			component: dialogComponent,
			preserveState: shouldPreserveState,
		})
	}

	// Triggers a context propagation, needed after the views were swapped,
	// so their properties can properly be updated accordingly.
	setContext()

	if (!shouldPreserveScroll) {
		resetScrollPositions()
	} else {
		restoreScrollPositions()
	}

	triggerEvent('navigate', options)
}

/** Initializes the router by reading the context and registering events if necessary. */
async function initializeRouter(): Promise<InternalRouterContext> {
	const context = getRouterContext()

	if (isBackForwardVisit()) {
		handleBackForwardVisit()
	} else if (isExternalVisit()) {
		handleExternalVisit()
	} else {
		debug.router('Handling the initial page visit.')

		// If we navigated to somewhere with a hash, we need to update the context
		// to add said hash because it was initialized without it.
		setContext({
			url: makeUrl(context.url, { hash: window.location.hash }).toString(),
		})

		await navigate({
			preserveState: true,
			replace: sameUrls(context.url, window.location.href),
		})
	}

	registerEventListeners()

	return context
}

/** Performs a local visit to the given component without a round-trip. */
async function performLocalComponentVisit(targetUrl: UrlResolvable, options: LocalVisitOptions) {
	const context = getRouterContext()
	const url = normalizeUrl(targetUrl)

	return await navigate({
		...options,
		payload: {
			version: context.version,
			dialog: context.dialog,
			url,
			view: {
				name: options.component ?? context.view.name,
				properties: options.properties,
			},
		},
	})
}

/** Performs an external visit by changing the URL directly. */
function performLocalExternalVisit(url: UrlResolvable, data?: VisitOptions['data']) {
	document.location.href = makeUrl(url, {
		search: qs.stringify(data, {
			encodeValuesOnly: true,
			arrayFormat: 'brackets',
		}),
	}).toString()
}
