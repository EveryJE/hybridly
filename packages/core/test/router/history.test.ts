import { afterAll, beforeAll, describe, expect, it, vi } from 'vitest'
import { setHistoryState } from '../../src/router/history'
import { fakeRouterContext, returnsArgs } from '../utils'

describe('setHistoryState', () => {
	beforeAll(() => {
		window.location.href = 'https://localhost'
	})

	afterAll(() => {
		vi.clearAllMocks()
	})

	it('pushes the given context into the history', () => {
		const spy = vi.spyOn(window.history, 'pushState').mockImplementation(returnsArgs)

		setHistoryState(fakeRouterContext())

		expect(spy).toHaveBeenCalledOnce()
		expect(spy.mock.results).toMatchSnapshot('context')
	})

	it('replaces the given context in the current history state', () => {
		const spy = vi.spyOn(window.history, 'replaceState').mockImplementation(returnsArgs)

		setHistoryState(fakeRouterContext(), { replace: true })

		expect(spy).toHaveBeenCalledOnce()
		expect(spy.mock.results).toMatchSnapshot('context')
	})
})
