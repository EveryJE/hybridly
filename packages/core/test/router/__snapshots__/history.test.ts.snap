// Vitest Snapshot v1

exports[`setHistoryState > pushes the given context into the history > context 1`] = `
[
  {
    "type": "return",
    "value": [
      {
        "dialog": undefined,
        "scrollRegions": [],
        "state": {},
        "url": "https://localhost/",
        "version": "abc123",
        "view": {
          "name": "default.view",
          "properties": {},
        },
      },
      "",
      "https://localhost/",
    ],
  },
]
`;

exports[`setHistoryState > replaces the given context in the current history state > context 1`] = `
[
  {
    "type": "return",
    "value": [
      {
        "dialog": undefined,
        "scrollRegions": [],
        "state": {},
        "url": "https://localhost/",
        "version": "abc123",
        "view": {
          "name": "default.view",
          "properties": {},
        },
      },
      "",
      "https://localhost/",
    ],
  },
]
`;
